---
title: "LD decay"
author: "Alejandro Thérèse Navarro"
date: "October 14th 2019"
output: html_document
---

Linkage Disequilibrium (LD) measures the co-inheritance of two alleles at two markers in the genome. We say two markers are in LD when the observed frequency of their alleles is different than the product of the frequencies of each allele. That is, with alleles a1 and b1 in loci **a** and **b**, if freq(a1)*freq(b1) = freq(a1b1) then the two loci are in linkage equilibrium. When the equality does not hold, we say that the loci are in LD. There are multiple causes of LD, the main one being physical linkage. For the purpose of this report, let us consider that physical linkage is the only cause of LD.

In general, as distance between markers increases, LD decreases, until a threshold distance is reached at which two markers are not in LD anymore. The concept of LD decay refers to this change of LD as distance increases. Ultimately, LD decay indicates to which extent are markers linked to their nearby regions. A fast LD decay means that high marker density is needed along the genome, while slow LD decay indicates that markers tend to be highly correlated. It is important to notice that LD is a **population measure** and not individual measure, and thus reflects population properties.

Measurement of LD is complex and non-uniform across literature, as Vos et al. (2017) explain in their article "Evaluation of LD decay and various LD-decay estimators in simulated and SNP-array data of tetraploid potato".They used different methods to calculate LD in tetraploids under different simulations. They propose a method to calculate LD based on the 90% percentile of LD at each distance window. This method has been applied to the NAM3 population we have been working with. 

To summarize, their method proposes to study the LD halflife. Let us see what this means. Short-range LD goes from high values at very short distances (for instance 0.5 r2 at 0.1 cM) to a "background LD" that is more or less constant accross large distances. LD takes x cM to go from its maximum to the level background correlation. The LD halflife is the point at which half of the decrease has passed. To find out this point, Vos et al. propose to fit a spline to the LD decay trend, and use that spline to estimate the halflife point. 

To achieve that, new functions have been developed. 

```{r functions,echo=F}

LD_decay <- function(dos,map,win_size = 0.1,max_dist = NULL,per_chr = F){

  #First we split dosages per chromosome and calculate correlations
  #only within chromosomes
  dos_per_chr <- split(dos,map$chromosome)
  LD <- lapply(dos_per_chr,function(g){
    ld <- cor(t(g))
    not_dup <- lower.tri(ld,diag=F)
    return(ld[not_dup]^2)
  })

  #then we calculate the distance between markers
  pos_per_chr <- split(map$position,map$chromosome)
  dis <- lapply(pos_per_chr,function(d){
    res <- sapply(d,function(p) abs(p-d))
    not_dup <- lower.tri(res,diag=F)
    return(res[not_dup])
  })

  if(!per_chr){

    #Per window we calculate the percentile estimates
    LD <- do.call(c,LD)
    dis <- do.call(c,dis)

    if(is.null(max_dist)) max_dist <- max(dis)

    windows <- seq(0,max_dist,win_size)
    ld_estimates <- t(sapply(seq_along(windows),function(w){
      sel <- dis >= windows[w] & dis < windows[w] + win_size
      return(quantile(LD[sel],c(0.5,0.8,0.9,0.95)))
    }))

    #We calculate the background correlation between chromosomes
    dos_sample <- lapply(dos_per_chr,function(d) d[sample(1:nrow(d),100),] )
    dos_sample <- do.call(rbind,dos_sample)
    back_ld <- cor(t(dos_sample))^2
    back_ld <- back_ld[lower.tri(back_ld,diag=F)]
    back_ld <- quantile(back_ld,c(0.5,0.8,0.9,0.95))

    #We add some extra features
    res <- list(LD = data.frame(ld_estimates,
                                distance = windows),
                background = back_ld)
    attr(res,"max_dist") <- max(dis,na.rm=T)
    attr(res,"class") <- c("LD","list")

  }else{
    res <- lapply(1:length(unique(map$chromosome)),function(i){
      ld <- LD[[i]]
      d <- dis[[i]]

      windows <- seq(0,max(d,na.rm=T),win_size)
      ld_estimates <- t(sapply(seq_along(windows),function(w){
        sel <- d > windows[w] & d < windows[w] + win_size
        return(quantile(ld[sel],c(0.5,0.8,0.9,0.95)))
      }))

      back_ld <- quantile(ld[d > max(d,na.rm=T)*0.9],c(0.5,0.8,0.9,0.95))

      #We add some extra features
      res <- list(LD = data.frame(ld_estimates,distance = windows),
                  background = back_ld)
      attr(res,"max_dist") <- max(d,na.rm=T)
      attr(res,"class") <- c("LD","list")
      return(res)
    })

  }

  return(res)
}

plot.LD <- function(LD,max_dist = NULL,main=NULL){
  if(is.null(max_dist)) max_dist <- attr(LD,"max_dist")
  if(max_dist > attr(LD,"max_dist"))
    stop("Cannot choose a larger max_dist than ",attr(LD,"max_dist"))
  if(!"LD" %in% class(LD)){
    stop("Object provided is not an LD list")
  }

  linkage <- LD$LD

  plot(0,type="n",
       xlab = "cM",ylab = "r2",main = main,
       xlim = c(0,max_dist),ylim=c(0,0.5))

  cols <- colorspace::qualitative_hcl(4)
  halflife <- c()
  for(i in 1:4){
    ld <- linkage[,i][linkage$distance <= max_dist]
    dis <- linkage$distance[linkage$distance <= max_dist]

    #We take out missing percentile values because spline cannot handle it
    if(any(is.na(ld))){
      ld <- na.omit(ld)
      dis <- dis[-attr(ld,"na.action")]
    }

    points(dis,ld,
           pch=19,cex=0.85,col=cols[i])

    sp <- smooth.spline(dis,ld,spar = 0.6)
    lines(sp,lwd=2,col= lighten(cols[i]))

    top <- max(ld,na.rm=T)
    halfpoint <- (top - LD$background[i])/2 + LD$background[i]
    abline(v = sp$x[which.min(abs(sp$y - halfpoint))],
           col=cols[i],
           lty=2,lwd=1.5)
    halflife[i] <- round(sp$x[which.min(abs(sp$y - halfpoint))],1)

  }
  legend("topright",pch=19,
         legend=paste(c("50th","80th","90th","95th")," ld1/2=",halflife)
         ,col=cols,bty="n")
  mtext(paste("background:",
              paste(c(names(LD$background)),
              round(LD$background,2),
              collapse=" ")))
}

```

Let's see what is the LD at a NAM3 population.
```{r,fig.width= 10,fig.height=8}
setwd("D:/Science/PhD/Code/mpQTL/mpQTL/")
source("R/viz_fun.R")
dos <- data.table::fread(
  "research/PedigreeSim/NAM_crosses/3_ancestral/cross200_alleledose.dat",
  header = T)[,-1]
map <- read.table("research/PedigreeSim/Potato.map",header=T)

LD <- LD_decay(dos,map,win_size = 0.5)
plot(LD,main="NAM3 LD decay",max_dist = 50)

```

The results show the LD calculated using the 50%, 80%, 90% and 95% percentiles (each line). This serves as an indication of the minimums and maximums of LD. As it can be seen, at short distances (< 5cM) there are markers with very low correlation, indicating repulsion-phase markers. In the legend, the "ld1/2" measure represents the halflife of each LD decay curve. As we can see, this halflife is very similar between all measures, indicating its robustness, and confirming Vos et al. report. 

```{r,fig.width= 10,fig.height=8}
plot(LD,main="closeup of LD",max_dist = 5)
```

Calculations per-chromosome have also been implemented. Although in this simulated population such calculations are not going to be very different per chromosome (although chromosome size affects LD), in real populations it is possible that there are large differences between chromosomes.
```{r,fig.width= 10,fig.height=8}
LD_perc <- LD_decay(dos,map,win_size = 0.5,per_chr = T)
for(i in 1:length(LD_perc)){
  plot(LD_perc[[i]],main=paste("chromosome",i),max_dist = 50)
}
```


