---
title: "QTL map viz"
author: "Alejandro Thérèse Navarro"
date: "4 de diciembre de 2018"
output: html_document
---

```{r}
#Ancestral
for(i in 1:length(data)){
  wald<-sapply(data[[i]]$result, function(r) r$wald)
  pval<-sapply(data[[i]]$result, function(r) r$pval)
  beta<-sapply(data[[i]]$result, function(r) length(r$beta))
  
  
  plot(-log10(pval),pch=19,cex=0.7,main=paste("Manplot",names(data)[i]))
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
  colQQplot(pval,main=paste("QQplot pval",names(data)[i]))
  
  plot(wald,main=paste("Wald test",names(data)[i]),pch=19,cex=0.7,col="red")
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
}
```

```{r}
#ancestral 2, with higher variance within ancestral group
for(i in 1:length(data)){
  wald<-sapply(data[[i]]$result, function(r) r$wald)
  pval<-sapply(data[[i]]$result, function(r) r$pval)
  beta<-sapply(data[[i]]$result, function(r) length(r$beta))
  
  
  plot(-log10(pval),pch=19,cex=0.7,main=paste("Manplot",names(data)[i]))
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
  colQQplot(pval,main=paste("QQplot pval",names(data)[i]))
  
  plot(wald,main=paste("Wald test",names(data)[i]),pch=19,cex=0.7,col="red")
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
}
```

```{r}
#Biallelic
for(i in 1:length(data)){
  wald<-sapply(data[[i]]$result, function(r) r$wald)
  pval<-sapply(data[[i]]$result, function(r) r$pval)
  beta<-sapply(data[[i]]$result, function(r) length(r$beta))
  
  
  plot(-log10(pval),pch=19,cex=0.7,main=paste("Manplot",names(data)[i]))
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
  colQQplot(pval,main=paste("QQplot pval",names(data)[i]))
  
  plot(wald,main=paste("Wald test",names(data)[i]),pch=19,cex=0.7,col="red")
  abline(v=data[[i]]$pos$QTL)
  abline(v=data[[i]]$pos$polyg,col="green",lty=2)
}
```

```{r}
pops<-c(0:9,rep(1:9,each=50))
for(i in 1:4){
      boxplot(data[[i]]$pheno[-1:-10]~pops[-1:-10],
            ylim=c(min(as.numeric(data[[i]]$pheno)),max(as.numeric(data[[i]]$pheno))),
            xlab="Cross index",ylab="phenotypes",main="Cross phenotypes")
    points(pops[2:10],data[[i]]$pheno[2:10],cex=1.5,pch=19)
    abline(h=data[[i]]$pheno[1],lty=2)
    text(9.5,data[[i]]$pheno[1],"central parent",xpd=T,cex=0.7)  
}

```

```{r}
Xs<-lapply(1:4,function(z){
  genotypes<-data[[z]]$geno
  lapply(1:3,function(i) dosage.X(as.matrix(genotypes[QTL[i],]),ploidy=4,normalize=F) )
})
QTL<-c(1348,4023,892)
for(z in 1:4){
  space <- 2
  y <- data[[z]]$pheno
  for (i in 1:3) {
    X<-Xs[[z]][[i]]
    plot(0, type = "n", 
         xlim = c(0, ncol(X) * space),
         ylim = c(min(y), max(y)),
         main=paste(names(data)[z],"QTL=",QTL[i]),
         xlab="alleles",ylab="phenotype",axes=F)
    axis(1,at=seq(0,ncol(X)*space,length.out=ncol(X))+0.5,
         labels = colnames(X),cex=0.7
         )
    axis(2)
    
    al<-as.numeric(colnames(X))
    anc<-matrix(0:399,ncol=10)
    anc.mat<-sapply(al,function(k){
    apply(anc,2,function(a) k%in%a)
    })
    
    an<-apply(anc.mat,2,which)
    an<-sapply(an,function(a) which(a==unique(an)))
    col <- viridis::viridis(length(unique(an)))
  
    for (i in 1:ncol(X)) {
      x <- X[, i] / 4 + (i - 1) * space
      no<-X[,i]!=0
      points(x[no],y[no],
        pch = 19,
        cex = 0.7,
        col = col[an[i]])
    }
  }
  
}

```


```{r}
pv<-sapply(result[[1]],function(k) k$pval )

plot(-log10(pv),pch=19,cex=0.7,main=paste("Manplot",names(data)[z]))
abline(v=data[[i]]$pos$QTL)
abline(v=data[[i]]$pos$polyg,col="green",lty=2)
colQQplot(pval,main=paste("QQplot pval",names(data)[z]))

pops<-c(0:9,rep(1:9,each=50))[sel]
      boxplot(phenotypes[-1:-4]~pops[-1:-4],
            ylim=c(min(phenotypes),max(phenotypes)),
            xlab="Cross index",ylab="phenotypes",main="Cross phenotypes")
    points(pops[2:4]-2,phenotypes[2:4],cex=1.5,pch=19)
    abline(h=phenotypes[1],lty=2)
    text(9.5,phenotypes[1],"central parent",xpd=T,cex=0.7)  
```

```{r}
#### Phenotype Variations ####
# In this section we will generate phenotypes using different
# parameters and then evaluate its effect on QTL detection
A3<-readRDS("data_A3.RDS")

#Positions of QTLs and polygenic effects
QTL<-c(1348,4023,892)
set.seed(7)
polyg<-sample(1:nrow(A3$geno),50)

#Genotypes (per chromosome) of the QTLs and polygenic effects
#the genotypes at the QTL positions
Qg<-A3$geno[QTL,] 
pg<-A3$geno[polyg,]
#Genetic map
map<-data.table::fread("../mpQTL/PedigreeSIM/Potato/Potato.map")
  
agdiversity<-c(0.01,0.1,1)
pgsize<-c(0.1,1,10)


pheno<-lapply(1:length(pgsize),function(f){
  #mu should be zero, sd depends on main effects (less than 1 is better).
  effects.polyg<-apply(pg,1,function(gen){
    e<-rnorm(length(unique(gen)),0,pgsize[f]) 
    names(e)<-unique(gen)
    return(e)
  })
  

  pheee<-lapply(1:length(agdiversity),function(w){
    #Obtain a set of QTL effects
    effects.QTL <- multi.effects2(
      gen = Qg,
      anc_alleles = matrix(0:399, nrow = 40),
      seed = 7,
      anc_sd = agdiversity[w]
    )
    
    phenotypes<-pheno2(genotypes = Qg,
                       effects = effects.QTL,
                       polygen=pg,
                       polygen.effects = effects.polyg,
                       ploidy=4,
                       herit = 0.8,
                       mu = 100,
                       return.effects = T)
    return(phenotypes)
  
    
  })
  names(pheee)<-agdiversity
  return(pheee)
})
names(pheno)<-pgsize

data<-list()
for(i in 1:length(pheno)){
  names(pheno)[i]
  for(j in 1:length(pheno[[i]])){
    data[[(i-1)*3+j]]<-pheno[[i]][[j]]
  }
}
#pgsize= 0.1, 0.1, 0.1, 1, 1, 1, 10, 10, 10
#agdiversity= 0.01, 0.1, 1, 0.01, 0.1, 1, 0.01, 0.1, 1
```

```{r}
res<-lapply(1:length(data),function(i){
  p<-data[[i]]$pheno
  r<-map.QTL(phenotypes = p,
          genotypes = A3$geno,
          map = map,
          dosage = A3$dosage,
          K = T,Q = T )
  return(r)
})
saveRDS(res,"results_diversity_test.RDS")

```


```{r}
res<-readRDS("results_diversity_test.RDS")
efs<-sapply(4:6,function(q) data[[q]]$effects[[1]])
cols<-rev(colorspace::sequential_hcl(3,h=120,c=100))
plot(0,type="n",axes=F,xlab="",ylab="",xlim=c(0,3),ylim=range(efs),
     main="QTL effects under different diversity scenarios")
axis(1,0:3,labels=paste("Diversity",1:3),at=c(0.5,1.5,2.5))
axis(2,round(seq(min(efs),max(efs),length.out = 10),1))

for(q in 1:3){
  
  ef<-pheno[[1]][[q]]$effects[[1]]
  
  anc<-matrix(0:399,ncol=10)
  anc.mat<-sapply(names(ef),function(k){
    apply(anc,2,function(a) k%in%a)
  })
  an<-apply(anc.mat,2,which)
  an<-sapply(an,function(a) which(a==unique(an)))
  
  points(an/4+1*(q-1),ef,col=cols[q],pch=19,cex=1.2,xlim=c(0,6))
}
```

```{r}

cols<-colorspace::sequential_hcl(3,h=120,c=100)
plot(0,type="n",axes=F,xlab="",ylab="",xlim=c(0,3),ylim=c(70,120),
     main="QTL effects under different diversity scenarios")
axis(1,0:3,labels=paste("Diversity",1:3),at=c(0.5,1.5,2.5))
axis(2,seq(min(data[[1]]$pheno,max(data[[1]]$pheno),length.out=10)))

for(q in 1:3){
  ef<-data[[q]]$pheno
  ancestral<-c(rep(NA,10),rep(1,100),rep(2,150),rep(3,200)) 
  points(ancestral/6+1*(q-1),ef,col=cols[q],pch=19,cex=0.7,xlim=c(0,6))
}

for(q in 7:9){
  ef<-data[[q]]$pheno
  ancestral<-c(c(1,1,1,2,2,2,3,3,3,3),rep(1,100),rep(2,150),rep(3,200))
  family<-c(c(0:9),rep(1:9,each=50))
  print(anova(aov(ef~ancestral)))
  print(anova(aov(ef~family)))
}
q<-8
anova(aov(ef~ancestral))
```

```{r}
cols<-rev(colorspace::sequential_hcl(3,h=100,c=120))
for(j in 1:3){
  opar<-par(no.readonly = T)
  r<-res[1:3+(j-1)*3]
  plot(0,type="n",
       xlim=c(0,length(r[[1]])),
       ylim=c(0,max(-log10(unlist(r)))),
       main="Manplot",xlab="markers",ylab="-log10(pval)"
       )
  abline(v=QTL)
  abline(v=polyg,col="grey",lty=2)
  for(i in 1:3){
    points(-log10(r[[i]]),pch=19,cex=0.5,main=paste("Manplot",3),col=cols[i])
  }
  comp.QQplot(do.call(cbind,r),coltype = "sequential",legend=paste("Div",1:3),main="QQplot")
  par(opar)
}

```

