---
title: "New phenotype simulation"
author: "Alejandro Thérèse Navarro"
date: "13 de noviembre de 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# source("mpQTL_fun.R")
cfiles<-paste0("../mpQTL/PedigreeSIM/NAM_crosses/",c(1,3,7,10),
               "_ancestral/cross",c(0,2,6,9),"00_founderalleles.dat")

phe.poly<-list()
phe.non<-list()
for(w in cfiles){
  genotypes<-link_NAM(crossfile=w,
                    totallele="../mpQTL/PedigreeSIM/Parents/Total_pop.txt")
  #main QTLs
  QTL<-genotypes[c(1348,4023,892),]
  effects.QTL<-multi.effects2(gen=QTL,
                              anc_alleles=matrix(0:399,nrow=40),
                              seed=7,
                              anc_sd=1)
  
  set.seed(7)
  polygenic<-sample(1:nrow(genotypes),50)
  polygen<-genotypes[polygenic,]
  
  effects.polyg<-apply(polygen,1,function(gen){
    e<-rnorm(length(unique(gen)),0,0.1) #mu should be zero, sd depends on main effects (anc_range).
    names(e)<-unique(gen)
    return(e)
  })
    
  phenotypes<-pheno2(genotypes = rbind(QTL,polygen),
         effects = c(effects.QTL,effects.polyg),
         ploidy=4,
         herit = 0.8,
         mu = 100)
  phe.poly[[w]]<-phenotypes
  phe.non[[w]]<-pheno2(genotypes = rbind(QTL),
         effects = c(effects.QTL),
         ploidy=4,
         herit = 0.8,
         mu = 100)
  # pheno2(QTL,effects,)
  # phenotype<-multi.pheno(tidy_allele(QTL),effects,herit = 0.5,mu = 200)
  # 
}
```
```{r}
pops<-c(0:9,rep(1:9,each=50))
for(i in 1:4){
      boxplot(as.numeric(phe.poly[[i]][-1:-10])~pops[-1:-10],
            ylim=c(min(as.numeric(phe.poly[[i]])),max(as.numeric(phe.poly[[i]]))),
            xlab="Cross index",ylab="phenotypes",main="Cross phenotypes")
    points(pops[2:10],phe.poly[[i]][2:10],cex=1.5,pch=19)
    abline(h=phe.poly[[i]][1],lty=2)
    text(9.5,phe.poly[[i]][1],"central parent",xpd=T,cex=0.7)  
}
```

```{r}
for(i in 1:4) {
  genotypes <- link_NAM(crossfile = cfiles[i],
  totallele = "PedigreeSIM/Parents/Total_pop.txt")
  #main QTLs
  QTL <- genotypes[c(1348, 4023, 892), ]
  effects.QTL <- multi.effects2(
                gen = QTL,
                anc_alleles = matrix(0:399, ncol= 10),
                seed = 7,
                anc_sd = 1,
                anc_range = c(0, 50)
                )
  
  for (j in 1:length(effects.QTL)) {
    anc <- apply(matrix(0:399, ncol = 10), 2, function(a) {
      res <- unique(QTL[j,]) %in% a
      names(res) <- unique(QTL[j,])
      return(res)
      })
    anc <- apply(anc, 1, which)
    anc <- anc[names(effects.QTL[[j]])]
    plot(effects.QTL[[j]], col = anc, pch = 19)
  }
}
```


```{r}
cfiles<-paste0("PedigreeSIM/NAM_crosses/",c(1,3,7,10),
               "_ancestral/cross",c(0,2,6,9),"00_founderalleles.dat")
dfiles<-paste0("PedigreeSIM/NAM_crosses/",c(1,3,7,10),
               "_ancestral/cross",c(0,2,6,9),"00_alleledose.dat")
mapfile<-"PedigreeSIM/Potato/Potato.map"
map<-data.table::fread(mapfile)

pval<-lapply(1:4,function(i){
  genotypes<-link_NAM(cfiles[i],totallele="PedigreeSIM/Parents/Total_pop.txt")
  dosage<-data.table::fread(dfiles[i])[,-1]
  ph<-phe.poly[[i]]
  pval<-map.QTL(phenotypes = ph,
          genotypes = genotypes,
          K=T,
          Q=T,k=5,
          dosage=dosage,
          map=map)
    pval2<-map.QTL(phenotypes = phe.non[[i]],
          genotypes = genotypes,
          K=T,
          Q=T,k=5,
          dosage=dosage,
          map=map)
  return(list(polygen=pval,main=pval2))
})

```

```{r}
for(i in 1:length(pval)){
  
  ymax<-max(-log10(unlist(pval[[i]])))
  
  plot(-log10(pval[[i]][[1]]),pch=19,cex=0.7, main="Red=non polygenic",ylim=c(0,ymax))
  abline(v=c(1348,4023,892))
  abline(v=polygenic,col="lightgreen",lty=2)
  
  points(-log10(pval[[i]][[2]]),pch=19,cex=0.5, col="red")
  
  plot(-log10(pval[[i]][[1]]),-log10(pval[[i]][[2]]),pch=19,cex=0.4,
       xlab="With polygenic effect",ylab="Without polygenic effect")
  abline(0,1,col="red",lwd=2)
}
```

```{r}
source("../../R/pheno_fun.R")
genotypes<-link_NAM(crossfile="PedigreeSIM/NAM_crosses/3_ancestral/cross200_founderalleles.dat",
                    totallele="PedigreeSIM/Parents/Total_pop.txt")
#main QTLs
QTL<-genotypes[c(1348,4023,892),]
set.seed(7)
polygenic<-sample(1:nrow(genotypes),50)
polygen<-genotypes[polygenic,]

effects.QTL<-multi.effects2(gen=QTL,
                            anc_alleles=matrix(0:399,nrow=40),
                            seed=7,
                            anc_sd=1, #difference within ancestral group
                            anc_range=c(0,50)) #range of difference between ancestral groups
effects.polyg<-apply(polygen,1,function(gen){
  e<-rnorm(length(unique(gen)),0,0.1) #mu should be zero, sd depends on main effects (anc_range).
  names(e)<-unique(gen)
  return(e)
})

pheno.gen<-phsum(QTL,effects.QTL,4,table=T)
pheno.poly<-phsum(polygen,effects.polyg,4)
env<-rnorm(460,sd=1)
mu<-rep(50,460)
pheno.table<-cbind(mu,pheno.gen,pheno.poly,env)
pheno.table<-cbind(pheno.table,rowSums(pheno.table))
colnames(pheno.table)<-c("mu","QTL1","QTL2","QTL3","polygen","env","total")
cov(pheno.table)


```

#### fun ---------

```{r, "fun"}
anc.finder<-function(QTL,anclist){
  if(is.vector(QTL)) QTL<-matrix(QTL,nrow=1)
  if(is.matrix(anclist)) anclist<-mat2list(anclist)
  result<-apply(QTL,1,function(k){
    sapply(anclist,function(a) any(k%in%a))
  })
  
  rownames(result)<-names(anclist)
  colnames(result)<-rownames(QTL)
  return(result) #+1-1 to transform logical into numeric
}

mat2list<-function(matrix,rowise=F){
  if(rowise){
    #if rowise we transpose the matrix so that rows are columns
    matrix<-t(matrix)
  }
  n<-ncol(matrix)
  result<-list()
  for(i in 1:n){
    result[[i]]<-matrix[,i]
  }
  return(result)
}

pheno2<-function(
  genotypes, #alleles per chromosome (loci x chromosome matrix) 
  effects, #as many effects as rows has the genotype matrix
  ploidy,
  seed=7,
  herit=0.5,
  mu=100,
  Evar=2
){
  phenotype<-phsum(genotypes,effects,ploidy)
  ######################################################### adapting phenotype to add variance
  set.seed(seed)
  env<-rnorm(length(phenotype),mean=mu,sd=Evar)
  Sg<-var(phenotype)
  Se<-var(env)
  
  #Now we will rescale the genetic effects
  alpha2<-Se*herit/(Sg*(1-herit))#calculation of alpha squared
  effects2<-lapply(effects,function(x) x*sqrt(alpha2))#rescaling of genetic effects
  #Calculation with new effects
  pheno2<-phsum(genotypes,effects2,ploidy)
  
  result<-pheno2+env#add up genetic and environmental effects
  return(result)
}

phsum<-function(
  genotypes,
  effects,
  ploidy,
  table=F #whether to give the results in table format (one for each locus) or already summed
  ){
  phenolist<-lapply(1:nrow(genotypes),function(k){ 
    gen<-genotypes[k,]
    ef<-effects[[k]]
    
    #incidence matrix (allele per chromosome)
    am<-sapply(gen,function(g) g==unique(gen)) 
    rownames(am)<-unique(gen)
    
    #reorder to match the order of the effects (must have same names)
    am<-t(am)[ ,names(ef)]
    cef<-am%*%ef #chromosome effect vector
    
    #Calculate phenotype at this locus (sum chromosomes per individual)
    ph<-sapply(1:(length(gen)/ploidy), function(i){
      j<-1:ploidy+(i-1)*ploidy
      r<-sum(cef[j]) #additive model (sums effects at each chromosome)
      
      names(r)<-paste(Reduce(
        #find the common characters in the names of each group of chromosomes
        intersect2, strsplit(names(cef[j,]), NULL)), 
        collapse = '')
      
      return(r)
    })
    return(ph)
  })
  if(table){
      return(Reduce(cbind,phenolist))
  }else{
      return(Reduce(`+`,phenolist))
  }

}

multi.effects2<-function(
  gen, #genotypes (per chromosome of individual)
  anc_alleles, #matrix (per column)/list defining which alleles belong to which ancestral group
  anc_range=c(0,10),#max and minimum ancestral mean
  anc_sd=1,
  seed=7
){
  
  if(is.matrix(anc_alleles)) anc_alleles<-mat2list(anc_alleles)
  #Identifying present ancestral groups (not useful for function)

    apply(gen, 1, function(g) {
      sapply(anc_alleles, function(i) {
      any(g %in% i)
      })
    })
  
  effects<-lapply(1:nrow(gen),function(k){
    #unique alleles present in locus k
    a<-unique(gen[k,]) 
    
    #to which ancestral does each allele correspond
    anc_matrix<-sapply(anc_alleles,function(c) a%in%c)
    rownames(anc_matrix)<-a
    
    #Number of alleles present in which ancestral groups
    anc_n<-colSums(anc_matrix)[colSums(anc_matrix)!=0]
    
    set.seed(seed*k)
    anc_mean<-runif(length(anc_n),min=anc_range[1],max=anc_range[2])
    
    eff<-lapply(1:length(anc_n),function(i){
      set.seed(seed*k*i)
      #randomly generate effects with a specific seed
      e<-rnorm(anc_n[i],anc_mean[i],sd=anc_sd)
      #which group are we taking (to put the right allele names)
      group<-which(colSums(anc_matrix)!=0)[i]
      names(e)<-rownames(anc_matrix)[which(anc_matrix[,group])]
      return(e)
    })
    names(eff)<-names(anc_n)
    
    return(do.call(c,eff))
  })

  return(effects)
}

intersect2<-function (x, y)
{
  y <- as.vector(y)
  y[match(as.vector(x), y, 0L)]
}


```

